crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+        OSSL_HPKE_err;
+        goto err;
+    }
+    kctx = EVP_KDF_CTX_new(kdf);
+    EVP_KDF_free(kdf); /* The kctx keeps a reference so this is safe */
+    kdf = NULL;
+    if (kctx == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Build up the parameters for the derivation */
+    if (mode5869 == OSSL_HPKE_5869_MODE_KEM) {
+        kem_ind = kem_iana2index(suite.kem_id);
+        if (kem_ind == 0) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        mdname = hpke_kem_tab[kem_ind].mdname;
+    } else {
+        kdf_ind = kdf_iana2index(suite.kdf_id);
+        if (kdf_ind == 0) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        mdname = hpke_kdf_tab[kdf_ind].mdname;
+    }
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST,
+                                            (char *)mdname, 0);
+    *p++ = OSSL_PARAM_construct_int(OSSL_KDF_PARAM_MODE, &mode);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_KEY,
+                                             (unsigned char *)labeled_ikm,
+                                             labeled_ikmlen);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT,
+                                             (unsigned char *)salt, saltlen);
+    *p = OSSL_PARAM_construct_end();
+    if (EVP_KDF_CTX_set_params(kctx, params) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    lsecretlen = EVP_KDF_CTX_get_kdf_size(kctx);
+    if (lsecretlen > *secretlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Do the derivation */
+    if (EVP_KDF_derive(kctx, secret, lsecretlen, params) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    *secretlen = lsecretlen;
+
+err:
+    EVP_KDF_free(kdf);
+    EVP_KDF_CTX_free(kctx);
+    memset(labeled_ikmbuf, 0, OSSL_HPKE_MAXSIZE);
+    return erv;
+}
+
+/*
+ * @brief RFC5869 HKDF-Expand
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param mode5869 - controls labelling specifics
+ * @param prk - the initial pseudo-random key material
+ * @param prk - length of above
+ * @param label - label to prepend to info
+ * @param labellen - label to prepend to info
+ * @param context - the info
+ * @param contextlen - length of above
+ * @param L - the length of the output desired
+ * @param out - the result of expansion (allocated by caller)
+ * @param outlen - buf size on input
+ * @return 1 for good otherwise bad
+ */
+static int hpke_expand(OSSL_LIB_CTX *libctx,
+                       const OSSL_HPKE_SUITE suite, const int mode5869,
+                       const unsigned char *prk, const size_t prklen,
+                       const char *label, const size_t labellen,
+                       const unsigned char *info, const size_t infolen,
+                       const uint32_t L,
+                       unsigned char *out, size_t *outlen)
+{
+    int erv = 1;
+    unsigned char libuf[INT_MAXSIZE];
```

Comment:
Again - WPACKET with a BUF_MEM is an option here.

Reply:
will respond on WPACKET above so marking this resolved